src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java

```diff
@@ -91,6 +167,156 @@ private boolean checkIfExecute()
+        {
+            return;
+        }
+
+        final AbstractBuildingWorker buildingWorker = getOwnBuilding();
+        if(buildingWorker instanceof BuildingBuilder)
+        {
+            ((BuildingBuilder) buildingWorker).resetNeededResources();
+        }
+
+        while (job.getStructure().findNextBlock())
+        {
+            @Nullable final Template.BlockInfo blockInfo = job.getStructure().getBlockInfo();
+            @Nullable final Template.EntityInfo entityInfo = job.getStructure().getEntityinfo();
+
+            if (entityInfo != null)
+            {
+                requestEntityToBuildingIfRequired(entityInfo);
+            }
+
+            if (blockInfo == null)
+            {
+                continue;
+            }
+
+            @Nullable IBlockState blockState = blockInfo.blockState;
+            @Nullable Block block = blockState.getBlock();
+
+            if (job.getStructure().isStructureBlockEqualWorldBlock()
+                    || (blockState.getBlock() instanceof BlockBed && blockState.getValue(BlockBed.PART).equals(BlockBed.EnumPartType.FOOT))
+                    || (blockState.getBlock() instanceof BlockDoor && blockState.getValue(BlockDoor.HALF).equals(BlockDoor.EnumDoorHalf.UPPER)))
+            {
+                continue;
+            }
+
+            if(block instanceof BlockSolidSubstitution)
+            {
+                blockState = getSolidSubstitution(job.getStructure().getBlockPosition());
+                block = blockState.getBlock();
+            }
+
+            final Block worldBlock = BlockPosUtil.getBlock(world, job.getStructure().getBlockPosition());
+
+            if (block != null
+                    && block != Blocks.AIR
+                    && worldBlock != Blocks.BEDROCK
+                    && !(worldBlock instanceof AbstractBlockHut)
+                    && !isBlockFree(block, 0))
+            {
+                requestBlockToBuildingIfRequired((BuildingBuilder) getOwnBuilding(), blockState);
+            }
+        }
+        job.getWorkOrder().setRequested(true);
+    }
+
+    /**
+     * Add blocks to the builder building if he needs it.
+     *
+     * @param building   the building.
+     * @param blockState the block to add.
+     */
+    private void requestBlockToBuildingIfRequired(BuildingBuilder building, IBlockState blockState)
+    {
+        if (job.getStructure().getBlockInfo().tileentityData != null)
+        {
+            final List<ItemStack> itemList = new ArrayList<>();
+            itemList.addAll(getItemsFromTileEntity());
+
+            for (final ItemStack stack : itemList)
+            {
+                building.addNeededResource(stack, 1);
+            }
+        }
+
+        building.addNeededResource(BlockUtils.getItemStackFromBlockState(blockState), 1);
+    }
+
+    /**
+     * Adds entities to the builder building if he needs it.
+     */
+    private void requestEntityToBuildingIfRequired(Template.EntityInfo entityInfo)
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 21 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)